fix type of rune and simple-rod for new SBCLs

slyrus opened the pull request:
The type inference has gotten stronger in recent SBCLs and the simple-rods might be either (simple-array rune) or simple-base-string. Accept either on SBCL.

slyrus committed:
fix type of rune and simple-rod for new SBCLs

scymtym commented:
Could you describe how to reproduce to problem?

slyrus commented:
The easiest way to see the problem is to do:

```
(asdf:test-system "xpath")
(xpath::run-all-tests)
````

This fails with the following error:

```
The value
  "div"
is not of type
  (SIMPLE-ARRAY CHARACTER)
when binding CXML::QNAME
   [Condition of type TYPE-ERROR]

Restarts:
 0: [RETRY] Retry SLIME REPL evaluation request.
 1: [*PROCESS-INPUT] Continue reading input.
 2: [ABORT] Return to SLIME's top level.
 3: [CLOSE-CONNECTION] Close SLIME connection.
 4: [ABORT] Exit debugger, returning to top level.

Backtrace:
  0: (CXML::SPLIT-QNAME "div") [external]
  1: (XPATH::DECODE-QNAME "div" #S(XPATH::DYNAMIC-ENVIRONMENT :NAMESPACES ((NIL . "") (NIL . "") ("xmlns" . "http://www.w3.org/2000/xmlns/") ("xml" . "http://www.w3.org/XML/1998/namespace"))) NIL)
  2: (XPATH::COMPILE-NODE-TEST "div" #S(XPATH::DYNAMIC-ENVIRONMENT :NAMESPACES ((NIL . "") (NIL . "") ("xmlns" . "http://www.w3.org/2000/xmlns/") ("xml" . "http://www.w3.org/XML/1998/namespace"))) NIL)
  3: (XPATH::COMPILE-LOCATION-STEP (:CHILD "div") #S(XPATH::DYNAMIC-ENVIRONMENT :NAMESPACES ((NIL . "") (NIL . "") ("xmlns" . "http://www.w3.org/2000/xmlns/") ("xml" . "http://www.w3.org/XML/1998/namespace..
  4: (XPATH::COMPILE-PATH ((:ROOT :NODE) (:CHILD "div")) #S(XPATH::DYNAMIC-ENVIRONMENT :NAMESPACES ((NIL . "") (NIL . "") ("xmlns" . "http://www.w3.org/2000/xmlns/") ("xml" . "http://www.w3.org/XML/1998/na..
  5: (XPATH:COMPILE-XPATH "/div" #S(XPATH::DYNAMIC-ENVIRONMENT :NAMESPACES ((NIL . "") (NIL . "") ("xmlns" . "http://www.w3.org/2000/xmlns/") ("xml" . "http://www.w3.org/XML/1998/namespace"))))
  6: (XPATH::TEST-FILTERING)
  7: (XPATH::RUN-ALL-TESTS)
  8: (SB-INT:SIMPLE-EVAL-IN-LEXENV (XPATH::RUN-ALL-TESTS) #<NULL-LEXENV>)
  9: (EVAL (XPATH::RUN-ALL-TESTS))
 --more--
```

Note that you have to explicitly call run-all-tests. Just using asdf:test-system doesn't actually call the tests. Using my branch allows the tests to run successfully.

slyrus commented:
More insight into the problem can be seen here:

```
XPATH-TEST> (type-of "/div")
(SIMPLE-ARRAY CHARACTER (4))

XPATH-TEST> (subtypep '(SIMPLE-ARRAY CHARACTER (4)) 'runes:simple-rod)
T
T

XPATH-TEST> (xpath::parse-xpath "/div")
(:PATH (:ROOT :NODE) (:CHILD "div"))

XPATH-TEST> (second (third (xpath::parse-xpath "/div")))
"div"

XPATH-TEST> (type-of (second (third (xpath::parse-xpath "/div"))))
(SIMPLE-BASE-STRING 3)

XPATH-TEST> (subtypep '(SIMPLE-BASE-STRING 3) 'runes:simple-rod)
NIL
T
```

So things start of with "/div" being a (SIMPLE-ARRAY CHARACTER (4)), which is a subtype of RUNES:SIMPLE-ROD. But as the query gets compiled, we get a string back which is (legitimately) a 
(SIMPLE-BASE-STRING 3) and, unfortunately, this is not a subtype of RUNES:SIMPLE-ROD. This change happened a few releases back in SBCL and seems to be a legitimate, and helpful, optimization, but closure-common can't deal with it (yet).

My proposed fix is:

```
-(deftype simple-rod () '(simple-array rune))
+(deftype simple-rod () #-sbcl '(simple-array rune) #+sbcl '(or (simple-array rune) simple-base-string))
```

scymtym commented:
Thank you for the example and explanation.

I tracked this down to the XPath lexer or more specifically the "fixup" lexer which post-processes the output of the main lexer. This is what happens:

1. The main lexer turns "div" into an `:ncname` token:
   ```lisp
   XPATH> (xpath-lexer "/div")
   #<CLOSURE (LAMBDA () :IN XPATH-LEXER) {100E8AA5AB}>
   XPATH> (funcall *)
   :/
   NIL
   XPATH> (funcall **)
   :NCNAME
   "div"
   ```
2. The fixup lexer turns "div" (as well as "mod", "and" and "or") into an operator name:
   ```lisp
   XPATH> (make-fixup-lexer (xpath-lexer "/div"))
   #<CLOSURE (LAMBDA () :IN MAKE-FIXUP-LEXER) {100E9DACAB}>
   XPATH> (funcall *)
   :/
   NIL
   XPATH> (funcall **)
   :DIV
   :DIV
   ```
3. The parser turns this kind of occurrence of the `:div` operator back into a string (via the `operator-actually-ncname` rule):
   ```
   XPATH> (second (third (xpath:parse-xpath "/div")))
   "div"
   XPATH> (type-of *)
   (SIMPLE-BASE-STRING 3)
   ```
   And this is where the `simple-base-string` comes from: SBCL stores symbol names as `simple-base-string` under certain circumstances and `operator-actually-ncname` returns the name of the keyword.

In light of this, it may be simpler and less invasive to adjust the parser instead of widening the `simple-rod` type. Something like
```diff
diff --git a/parser.lisp b/parser.lisp
index b0b550a..4d40783 100644
--- a/parser.lisp
+++ b/parser.lisp
@@ -115,8 +115,7 @@
    (:ns-name (lambda* (a) `(:namespace ,a)))
    (:qname (lambda* (a) `(:qname ,(car a) ,(cdr a))))
    :ncname
-   (operator-actually-ncname
-    (lambda* (a) (string-downcase (symbol-name a))))
+   operator-actually-ncname
    (:star (lambda* (nil) '*))
    (:node-type-or-function-name
     :lparen :rparen
@@ -127,7 +126,11 @@
    (:processing-instruction :lparen :literal :rparen
                             (lambda* (nil nil a nil)
                                      `(:processing-instruction ,a))))
-  (operator-actually-ncname :mod :div :and :or)
+  (operator-actually-ncname
+   (:mod (lambda* (a) (declare (ignore a)) "mod"))
+   (:div (lambda* (a) (declare (ignore a)) "div"))
+   (:and (lambda* (a) (declare (ignore a)) "and"))
+   (:or (lambda* (a) (declare (ignore a)) "or")))
 
   ;;;;;;;;;;;; 2.4 ;;;;;;;;;;;;
   ;;
```
seems to work.

What do you think?

slyrus commented:
This is an interesting idea but I have a couple of thoughts:
1) why not allow a rod to be a simple-base-string? Seems like it might be a nice thing for the (SBCL) compiler to treat these as simple-base-strings (faster code maybe?).

2) There are other ways to get a simple-base-string from the parser. I haven't tracked it down yet, but doing things like "/foo[baz]" returns foo as a simple-base-string. There may be was to track down those cases too but it seems simpler to just let rods be simple-base-strings.

slyrus commented:
It's not quite as simple (hah!) as I made it sound like above as:

```
(xpath:evaluate "/foo[baz]" *sample-xml-stp*)
```

works

but

```
(xpath:evaluate (format nil "/foo[baz]") *sample-xml-stp*)
```

doesn't. I stand by claim that a simple-base-string should be able to be a rod.

scymtym commented:
I agree that tracking down all instances in which the parser may produce `simple-base-string`s is probably difficult. If you don't object, I would still like to make the change I suggested above since it makes the parser less convoluted in my opinion. It should be orthogonal to the issue at hand.

Regarding the actual issue: if you are confident that changing the definition of `simple-rod` in the proposed way does not introduce problems, I'm OK with it.

slyrus commented:
Here's an excerpt from an email exchange with @lichtblau (I hope you don't my posting this here publicly, @lichtblau ):

```
> The specific issue is that at some point in the past
> year or two SBCL started making more strings be SIMPLE-BASE-STRINGs which
> are not (SIMPLE-ARRAY CHARACTER)s (or, in this case, (SIMPLE_ARRAY RUNE)).
> So, I modified the type of simple-rod as follows:
>
> -(deftype simple-rod () '(simple-array rune))
> +(deftype simple-rod () #-sbcl '(simple-array rune) #+sbcl '(or (simple-array rune) simple-base-string))

Apologies in advance that my advice is going not going to be awfully
useful, but it is the best I can offer based on the little that I
remember:

- Are we sure that this is not a pessimization?  The simple-rod type
  does not appear in a lot of places, but Gilbert's original code base
  was definitely optimized for speed and with the vague memories of Lisp
  compilers that I have, it seems hard to imagine that a union type
  would lead to string operations being inlined correctly.  It may
  matter -- or indeed it might not matter at all.  If code gets slow in
  the forest and nobody has a benchmark to notice, does it matter?

- In practical terms, the only part of the code base which might be fast
  is the original XML parser.  Going through STP, XPath, or XSLT will
  throw away all the speed anyway.  So one benchmark which may indicate
  whether any of this is of relevance could be to run the XML test
  suite, which parses a lot of different documents.  Or just to parse
  some other huge document into nowhere.

- One alternative could be to consider xpath::decode-qname, which, if I
  read this right, calls the unexported cxml::split-qname.  Is it the
  fault of the call site that it does not coerce data to the type
  expected by the callee?

- But I am having some difficulty justifying my own recommendation to
  myself, because the existing declaration looks like it is incorrect in
  the first place.  Should not the dimension rank be one-dimensional,
  i.e. shouldn't it be:

  -(deftype simple-rod () '(simple-array rune))
  +(deftype simple-rod () '(simple-array rune (*)))

  Somewhat hard to argue that the current code is optimized if has this
  mistake.  Can we fix that part at the same time?

- If either decode-qname itself or some helper function were improved to
  (coerce ... 'simple-rod), this bad type declaration would also hit,
  because coerce can deal with the fixed (*) but not the original *
  type.
```

This explains why just trying to coerce a simple-base-string to a 'runes::simple-rod didn't work -- the type definition for simple-rod was wrong! Fixing this allows us to just take the results of string-downcase and coerce that to a 'runes:simple-rod. I suggest that we fix the type defintion of 'runes:simple-rod, simplify your approach to just coercing the output of string-downcase symbol-name in operator-actually-ncname. That still leaves the issue of what to do about my xpath problem, but I think those can be tracked down without resorting to the union type -- although, as licthtblau points out, the type declaration probably wasn't doing much since it was wrong :)
